### PR TITLE
Download Failure Robustness/Recovery

### PR DESCRIPTION
**Pull request by rmartin16**
## Changes
- Download to a temporary file
  - Instead of downloading directly to the Briefcase cache, files are downloaded to the platform's designated temporary directory.
  - If the download succeeds, the temporary file is moved in to the Briefcase cache.
  - In most cases, the temporary file will be deleted is the download fails; one such exception may be Briefcase receiving SIGKILL.

## PR Checklist:

- [ ] All new features have been tested
- [ ] All new features have been documented
- [ ] I have read the **CONTRIBUTING.md** file
- [ ] I will abide by the code of conduct

### CODE REVIEW

File: src/briefcase/integrations/download.py [hunk 1]
```diff
@@ -97,3 +87,38 @@ def file(self, url, download_path, role=None):
...
+                    task_id = progress_bar.add_task("Downloader", total=int(total))
+                    with progress_bar:
+                        for data in response.iter_content(chunk_size=1024 * 1024):
+                            temp_file.write(data)
+                            progress_bar.update(task_id, advance=len(data))
+
+            self.tools.shutil.move(temp_file.name, filename)
+            # retrieving the current umask requires changing it...
+            os.umask(current_umask := os.umask(0o022))
+            self.tools.os.chmod(filename, 0o664 & ~current_umask)
```
Comment (rmartin16) [inline src/briefcase/integrations/download.py:119]: The `NamedTemporaryFile` file is only created with `600` permissions. So, all this `umask` stuff restores the previous behavior from `filename.open()`.

File: src/briefcase/integrations/download.py [hunk 2]
```diff
@@ -97,3 +87,38 @@ def file(self, url, download_path, role=None):
...
+                    temp_file.write(response.content)
+                else:
+                    progress_bar = self.tools.input.progress_bar()
+                    task_id = progress_bar.add_task("Downloader", total=int(total))
+                    with progress_bar:
+                        for data in response.iter_content(chunk_size=1024 * 1024):
+                            temp_file.write(data)
+                            progress_bar.update(task_id, advance=len(data))
+
+            self.tools.shutil.move(temp_file.name, filename)
```
Comment (rmartin16) [inline src/briefcase/integrations/download.py:116]: I just realized this `move` may not be a simple inode update if the system's temporary directory is on a filesystem different than that of Briefcase's cache directory. In that case, Python will fall back to a `copy` and `rm`.

Therefore, the underlying problem still exists if a user in this situation sends CTRL+C during the move potentially leaving an incomplete file in the Briefcase cache. (There's also the issue of the `move` potentially taking a relatively long time.)

Given this, I'll pause for feedback:
- Should I move forward with this approach? Or re-implement this solution always creating a temp file within the Briefcase cache?

The primary advantage IMO of the `NamedTemporaryFile` file is if anything goes wrong, the file already exists in an OS-designated temporary directory with (potential) OS faculties to delete it whenever.

I can also wrap the file move in a transient Wait Bar in case it is slow.

Comment (rmartin16) [reply src/briefcase/integrations/download.py:116]: hmm....another thought is if the Briefcase cache is on a different filesystem, that is likely because the default filesystem is low on space....and that same default filesystem is likely to be used by Python's `NamedTemporaryFile`. Therefore, using it for this purpose here would probably betray some users' configuration....guess I have my answer...

Comment (rmartin16) [reply src/briefcase/integrations/download.py:116]: on third thought....`NamedTemporaryFile` supports specifying a directory to create the temporary file in....so, that can mitigate the cross-filesystem concerns and still delegate creating a unique filename to Python's `tempfile`. We could even use the support to add a `suffix` or `prefix` to the filename to detect if one of these temporary files persists between Briefcase runs.

Comment (rmartin16) [reply src/briefcase/integrations/download.py:121]: ## tl;dr
Updated the creation of the temporary file to occur in the directory of the Briefcase cache that the file will ultimately be moved to. This mitigates the concerns here

File: src/briefcase/integrations/download.py [hunk 3]
```diff
@@ -97,3 +87,43 @@ def file(self, url, download_path, role=None):
...
+
+        :param response: ``requests.Response``
+        :param filename: full filesystem path to save data
+        """
+        temp_file = tempfile.NamedTemporaryFile(
+            dir=filename.parent,
+            prefix=f"{filename.name}.",
+            suffix=".download",
+            delete=False,
+        )
```
Comment (rmartin16) [inline src/briefcase/integrations/download.py:107]: By default, `tempfile` uses random chars for the filename; prefixing with the intended filename makes it possible to know what the temporary data is and the suffix makes it easy to identity it as a temporary file.

Here's an example filename:
```
Python-3.10-linux-x86_64-support.b4.tar.gz.vdqfwdt2.download
```

**Comment by rmartin16** [comment]
As mentioned in #753:

> The only thing to be wary of is the location of the temp file - if we just append ".tmp" to the filename in the final location, we could end up with stray partial download files that are difficult for users to find and are never cleaned up - or, even worse, we could end up with collisions on the temp file name. We either need to be very careful about doing pre/post download cleanup, or make use of temporary directories to ensure that there won't be collisions.

Right now, this PR doesn't address the possibility of `*.download` files accumulating in the Briefcase cache because of some bizarre unaccounted for problem.

So, should I also make updates to glob the Briefcase cache directory for `*.download` files at startup and delete them? Seems less than ideal...

Alternatively, perhaps this type of cleanup would be more appropriate as part of a "[a] user-accessible mechanism to manually flush the download cache".

File: tests/integrations/download/test_Download__file.py [hunk 4]
```diff
@@ -86,6 +89,11 @@ def test_new_download_oneshot(mock_tools, url, content_disposition):
     # The filename is derived from the URL or header
     assert filename == mock_tools.base_path / "downloads" / "something.zip"
 
+    # Temporary file was upgraded to intended destination
+    mock_tools.shutil.move.assert_called_with(mock.ANY, filename)
```
Comment (freakboy3742) [inline tests/integrations/download/test_Download__file.py:93]: While this test passes, it's not especially resilient - `mock.ANY` really does allow *any* content. This initial "mock.ANY" call is definitely required to verify that *a* filename was used, but it would be desirable to add checks that the same temporary filename was used for all the subsequent calls.

File: src/briefcase/integrations/download.py (hunk 1 above)
Comment (freakboy3742) [reply src/briefcase/integrations/download.py:124]: A longer explanatory comment would be helpful here. The interplay between the 0o022 mask and the `and not` with 0o644 wan't clear to me.

There's also no test that the resulting file mask is actually followed; there's code coverage, but no real check of the logic.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
I'm sure there are edge cases where this will still fail, but they're a lot smaller than the current edge case. 2 testing related comments, and one request for clarification in a docstring, but otherwise this looks really solid.

File: tests/integrations/download/test_Download__file.py (hunk 4 above)
Comment (rmartin16) [reply tests/integrations/download/test_Download__file.py]: That's fair; I was initially struggling with how to introspect information about the temporary file and I went too far towards "assume it's working if everything else is working." I think my changes now verify the logic is working as designed.

File: src/briefcase/integrations/download.py (hunk 1 above)
Comment (rmartin16) [reply src/briefcase/integrations/download.py]: Added (some probably overly verbose) comments about umask. I also tried to implement file permission verification in the tests.....things are really wonky on windows, though...

**Review by freakboy3742 — APPROVED** [review]
Awesome - nice work.

Also - Kudos on the first walrus in the codebase :-P
